src/GovUk.Education.ExploreEducationStatistics.Publisher/Functions/PublishReleaseFilesFunction.cs

```diff
@@ -47,75 +43,62 @@ public class PublishReleaseFilesFunction
-                        executionContext.FunctionName);
-                    logger.LogError("{StackTrace}", e.StackTrace);
+            await UpdateFilesStage(message.Releases, Started);
 
-                    await UpdateStage(releaseId, releaseStatusId, Failed,
-                        new ReleasePublishingStatusLogMessage($"Exception in files stage: {e.Message}"));
-                }
-            }
-
-            try
-            {
-                if (immediate)
-                {
-                    await _queueService.QueuePublishReleaseDataMessagesAsync(published);
-                }
-                else
+            var successfulReleases = await message
+                .Releases
+                .ToAsyncEnumerable()
+                .WhereAwait(async releaseStatus =>
                 {
-                    await _queueService.QueueGenerateReleaseContentMessageAsync(published);
-                }
+                    try
+                    {
+                        await _publishingService.PublishMethodologyFilesIfApplicableForRelease(releaseStatus.ReleaseId);
+                        await _publishingService.PublishReleaseFiles(releaseStatus.ReleaseId);
+                        return true;
+                    }
+                    catch (Exception e)
+                    {
+                        logger.LogError(e, "Exception occured while executing {FunctionName}",
+                            executionContext.FunctionName);
 
-                foreach (var (releaseId, releaseStatusId) in published)
-                {
-                    await UpdateStage(releaseId, releaseStatusId, Complete);
-                }
-            }
-            catch (Exception e)
-            {
-                logger.LogError(e, "Exception occured while executing {0}",
-                    executionContext.FunctionName);
-                logger.LogError("{0}", e.StackTrace);
-            }
+                        logger.LogError("{StackTrace}", e.StackTrace);
 
-            logger.LogInformation("{0} completed",
-                executionContext.FunctionName);
-        }
+                        return false;
+                    }
+                })
+                .ToListAsync();
 
-        private async Task<bool> IsImmediate(PublishReleaseFilesMessage message)
-        {
-            if (message.Releases.Count() > 1)
-            {
-                // If there's more than one Release this invocation couldn't have been triggered for immediate publishing
-                return false;
-            }
+            var unsuccessfulReleases = message
+                .Releases
+                .Except(successfulReleases);
+
+            await UpdateFilesStage(successfulReleases, Complete);
+            await UpdateFilesStage(unsuccessfulReleases, Failed);
+                
+            await _publishingCompletionService.CompletePublishingIfAllPriorStagesComplete(
+                successfulReleases, 
+                DateTime.UtcNow);
```

Comment:
Feels like this is something which can potentially fail and raise an exception, in which case the function host will retry this message up to the retry limit, effectively re-running the files stage multiple times.

I think I would do some testing around how well the code prior to this copes if it's re-run multiple times, and consider catching any exception that could be thrown here, or consider disabling retry functionality. I'm not sure if you can do this at a function level, so it might have to be done across the function project.

Reply:
By the looks of it, the prior code to do with file transfers all deals with *copying* folders from private to public storage rather than moving or copying-then-deleting.  By the looks of it then, it should be re-runnable.  

I can't see a reason though not to wrap in a try-catch.  In some ways it would be safer to iterate over the `successfulReleases` list and attempt to process each one in turn.  This has the advantage that individually there's more chance that a single Release in this list will be published successfully, as the failure of one in this list to publish might mean that all the others in this list would *also* fail, whereas iterating over each and publishing individually with a try-catch would get as many out as possible.

The downside of this approach is that we lose any potential optimisations we might have over determining which Publications need updating, superseding, based on a batch of Release Ids, so we might end up updating the same one several times over with this iteration / release-by-release approach.  Looking in the PulicationCompletionService, you can see that given a batch of Release Ids, we then determine a batch of Publication Ids that might need updating based upon them.  Be interested to hear your thoughts?

For now, I'm happy to wrap this in a try-catch.

Reply:
I've also amended the code to set the "Publishing" stage to failed for all of these Releases if this particular method fails.

Reply:
I agree that doing it by release would ultimately be better so that a failure copying files doesn't impact the publishing of releases in other publications but I think it's best that the optimisations in the completion step are retained for now by calling it once for a list of all the successful release id's, rather than doing it release by release because it's doing some tasks at a publication level which would get repeated.